### PR TITLE
Python version of xmlquery

### LINKED ISSUE #105: Rewrite xmlquery in python

### PR DESCRIPTION
**Pull request by wilke**
Changed xmlquery to python
Added get_values function to cases

Closes #105

### CODE REVIEW

File: utils/python/CIME/XML/entry_id.py [hunk 1]
```diff
@@ -86,6 +86,40 @@ def get_type_info(self, vid):
...
+    # Get description , expect child with tag "description" for parent node
+    def _get_description (self, node):
+        type_node = self.get_optional_node("desc", root=node)
+        if type_node is not None:
+            return type_node.text
+        else:
+            return None
+
+    # Get group , expect child with tag "group" for parent node        
+    def _get_group (self, node):
```
Comment (jedwards4b) [inline utils/python/CIME/XML/entry_id.py]: group is no longer a field within each variable, instead it looks like:
'''
  <group id="external_tools">
    <entry id="DATA_ASSIMILATION" value="FALSE">
      <type>logical</type>
      <valid_values>TRUE,FALSE</valid_values>
      <desc> Run the external tool pointed to by DATA_ASSIMILATION_SCRIPT after the model run completes </desc>
    </entry>
    <entry id="DATA_ASSIMILATION_CYCLES" value="1">
      <type>integer</type>
      <valid_values/>
      <desc> Number of model run - data assimilation steps to complete </desc>
    </entry>
    <entry id="DATA_ASSIMILATION_SCRIPT" value="">
      <type>char</type>
      <valid_values/>
      <desc>External script to be run after model completion</desc>
    </entry>
  </group>
'''

File: utils/python/CIME/XML/env_archive.py [hunk 2]
```diff
@@ -15,17 +15,102 @@ def __init__(self, case_root=None, infile="env_archive.xml"):
...
+                default = self._get_default(keep)
+                file    = self.filename
+            
+                #t   =  super(EnvBase , self).get_type( node )
+                v = { 'group' : group , 'attribute' : attr , 'value' : val , 'type' : t , 'description' : desc , 'default' : default , 'file' : file}
+                logger.debug("Found node with value for %s = %s" , item , v )   
+                results.append(v)
+            
+            
+            # val = { 'group' : None , 'attribute' : item , 'value' : node.text , 'type' : None , 'description' : None }
```
Comment (jedwards4b) [inline utils/python/CIME/XML/env_archive.py]: remove debug comments

File: utils/python/CIME/XML/env_batch.py [hunk 3]
```diff
@@ -54,8 +54,71 @@ def get_value(self, item, attribute={}, resolved=True, subgroup="run"):
...
+
         return value
 
+    def get_values(self, item, attribute={}, resolved=True, subgroup=None):
+        """Returns the value as a string of the first xml element with item as attribute value. 
+        <elememt_name attribute='attribute_value>value</element_name>"""
+        
+        logger.debug("(get_values) Input values: %s , %s , %s , %s , %s" , self, item, attribute, resolved, subgroup)
+        
+        logger.warning("In get_values (%s)" , self.__class__.__name__)
```
Comment (jedwards4b) [inline utils/python/CIME/XML/env_batch.py]: remove debug warning

File: utils/python/CIME/XML/generic_xml.py [hunk 4]
```diff
@@ -28,7 +30,10 @@ def __init__(self, infile=None):
         else:
             # if file does not exist create a root xml element
             # and set it's id to file
+
+            logger.warning("File %s does not exists." , infile)
```
Comment (jedwards4b) [inline utils/python/CIME/XML/generic_xml.py]: change this from warning to debug?

File: utils/python/CIME/case.py [hunk 5]
```diff
@@ -115,26 +123,78 @@ def get_value(self, item, attribute={}, resolved=True, subgroup=None):
         result = None
         for env_file in self._env_entryid_files:
             # Wait and resolve in self rather than in env_file
+
```
Comment (jedwards4b) [inline utils/python/CIME/case.py]: Instead of two loops here you can use for env_file in self._files

File: utils/python/CIME/case.py [hunk 6]
```diff
@@ -115,26 +123,78 @@ def get_value(self, item, attribute={}, resolved=True, subgroup=None):
...
+    
             if result is not None:
                 if resolved and type(result) is str:
                     return self.get_resolved_value(result)
                 return result
+       
+        # Return empty result
+        return result
 
-        result = None
```
Comment (jedwards4b) [inline utils/python/CIME/case.py]: I think that we still need this lookups dict

File: utils/python/CIME/case.py [hunk 7]
```diff
@@ -115,26 +123,78 @@ def get_value(self, item, attribute={}, resolved=True, subgroup=None):
...
-            type_str = self._get_type_info(node)
-            return convert_to_type(result, type_str, item)
+    def get_values(self, item=None, attribute={}, resolved=True, subgroup=None):
+        
+        """
+        Return info object for given item, return all info for all item if item is empty.     
+        """
+  
+        results = []
+        for env_file in self._env_entryid_files:
```
Comment (jedwards4b) [inline utils/python/CIME/case.py]: Can this also be a single loop using self._files?

File: utils/python/CIME/XML/entry_id.py (hunk 1 above)
Comment (wilke) [inline utils/python/CIME/XML/entry_id.py]: Changed

File: utils/python/CIME/XML/env_archive.py (hunk 2 above)
Comment (wilke) [inline utils/python/CIME/XML/env_archive.py]: removed from entry_id as well

File: utils/python/CIME/XML/generic_xml.py (hunk 4 above)
Comment (wilke) [inline utils/python/CIME/XML/generic_xml.py]: Should be a warning so if the file is missing a user knows that the case dir is incomplete and something is broken?

File: utils/python/CIME/case.py (hunk 5 above)
Comment (wilke) [inline utils/python/CIME/case.py]: changed

File: utils/python/CIME/case.py (hunk 6 above)
Comment (wilke) [inline utils/python/CIME/case.py]: added it back - but it seems lookups dict is empty

File: utils/python/CIME/case.py [hunk 8]
```diff
@@ -31,6 +31,7 @@
 
 from CIME.XML.generic_xml           import GenericXML
 import glob, shutil
+import traceback
```
Comment (jedwards4b) [inline utils/python/CIME/case.py]: For debug?

**Comment by jgfouca** [comment]
@wilke , what's the status of this PR?

**Comment by rljacob** [comment]
I was just talking to @wilke.  Tests need to be run on this branch.  Blues is still down today.

**Comment by jgfouca** [comment]
Please PLEASE don't do primary development on blues! :)

If you give me login access to a good desktop, I'll set it up for you!

**Pull request merged**